README.md
Beeswarm-style plots with ggplot2
================

[![Build
Status](https://travis-ci.org/eclarke/ggbeeswarm.svg?branch=master)](https://travis-ci.org/eclarke/ggbeeswarm)
[![CRAN
status](https://www.r-pkg.org/badges/version/ggbeeswarm)](https://cran.r-project.org/package=ggbeeswarm)

## Introduction

Beeswarm plots (aka column scatter plots or violin scatter plots) are a
way of plotting points that would ordinarily overlap so that they fall
next to each other instead. In addition to reducing overplotting, it
helps visualize the density of the data at each point (similar to a
violin plot), while still showing each data point individually.

`ggbeeswarm` provides two different methods to create beeswarm-style
plots using [ggplot2](http://ggplot2.org). It does this by adding two
new ggplot geom objects:

-   `geom_quasirandom`: Uses a [van der Corput
    sequence](http://en.wikipedia.org/wiki/Van_der_Corput_sequence) or
    Tukey texturing (Tukey and Tukey “Strips displaying empirical
    distributions: I. textured dot strips”) to space the dots to avoid
    overplotting. This uses
    [sherrillmix/vipor](https://github.com/sherrillmix/vipor).

-   `geom_beeswarm`: Uses the
    [beeswarm](https://cran.r-project.org/web/packages/beeswarm/index.html)
    library to do point-size based offset.

Features:

-   Can handle categorical variables on the y-axis (thanks @smsaladi,
    @koncina)
-   Automatically dodges if a grouping variable is categorical and
    `dodge.width` is specified (thanks @josesho)

See the examples below.

## Installation

This package is on CRAN so install should be a simple:

``` r
install.packages('ggbeeswarm')
```

If you want the development version from GitHub, you can do:

``` r
devtools::install_github("eclarke/ggbeeswarm")
```

## Examples

Here is a comparison between `geom_jitter` and `geom_quasirandom` on the
`iris` dataset:

``` r
set.seed(12345)
library(ggplot2)
library(ggbeeswarm)
#compare to jitter
ggplot(iris,aes(Species, Sepal.Length)) + geom_jitter()
```

<img src="README_files/figure-gfm/ggplot2-compare-1.png" width="576" />

``` r
ggplot(iris,aes(Species, Sepal.Length)) + geom_quasirandom()
```

<img src="README_files/figure-gfm/ggplot2-compare-2.png" width="576" />

### geom_quasirandom()

Using `geom_quasirandom`:

``` r
#default geom_quasirandom
ggplot(mpg,aes(class, hwy)) + geom_quasirandom()
```

<img src="README_files/figure-gfm/ggplot2-examples-1.png" width="576" />

``` r
# With categorical y-axis
ggplot(mpg,aes(hwy, class)) + geom_quasirandom(groupOnX=FALSE)
```

<img src="README_files/figure-gfm/ggplot2-examples-2.png" width="576" />

``` r
# Some groups may have only a few points. Use `varwidth=TRUE` to adjust width dynamically.
ggplot(mpg,aes(class, hwy)) + geom_quasirandom(varwidth = TRUE)
```

<img src="README_files/figure-gfm/ggplot2-examples-3.png" width="576" />

``` r
# Automatic dodging
sub_mpg <- mpg[mpg$class %in% c("midsize", "pickup", "suv"),]
ggplot(sub_mpg, aes(class, displ, color=factor(cyl))) + geom_quasirandom(dodge.width=1)
```

<img src="README_files/figure-gfm/ggplot2-examples-4.png" width="576" />

#### Alternative methods

`geom_quasirandom` can also use several other methods to distribute
points. For example:

``` r
ggplot(iris, aes(Species, Sepal.Length)) + geom_quasirandom(method = "tukey") + ggtitle("Tukey texture")
```

<img src="README_files/figure-gfm/ggplot2-methods-1.png" width="576" />

``` r
ggplot(iris, aes(Species, Sepal.Length)) + geom_quasirandom(method = "tukeyDense") +
    ggtitle("Tukey + density")
```

<img src="README_files/figure-gfm/ggplot2-methods-2.png" width="576" />

``` r
ggplot(iris, aes(Species, Sepal.Length)) + geom_quasirandom(method = "frowney") +
    ggtitle("Banded frowns")
```

<img src="README_files/figure-gfm/ggplot2-methods-3.png" width="576" />

``` r
ggplot(iris, aes(Species, Sepal.Length)) + geom_quasirandom(method = "smiley") +
    ggtitle("Banded smiles")
```

<img src="README_files/figure-gfm/ggplot2-methods-4.png" width="576" />

``` r
ggplot(iris, aes(Species, Sepal.Length)) + geom_quasirandom(method = "pseudorandom") +
    ggtitle("Jittered density")
```

<img src="README_files/figure-gfm/ggplot2-methods-5.png" width="576" />

``` r
ggplot(iris, aes(Species, Sepal.Length)) + geom_beeswarm() + ggtitle("Beeswarm")
```

<img src="README_files/figure-gfm/ggplot2-methods-6.png" width="576" />

### geom_beeswarm()

Using `geom_beeswarm`:

``` r
ggplot(iris,aes(Species, Sepal.Length)) + geom_beeswarm()
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-1.png" width="576" />

``` r
ggplot(iris,aes(Species, Sepal.Length)) + geom_beeswarm(side = 1L)
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-2.png" width="576" />

``` r
ggplot(mpg,aes(class, hwy)) + geom_beeswarm(size=.5)
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-3.png" width="576" />

``` r
# With categorical y-axis
ggplot(mpg,aes(hwy, class)) + geom_beeswarm(size=.5)
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-4.png" width="576" />

``` r
# Also watch out for points escaping from the plot with geom_beeswarm
ggplot(mpg,aes(hwy, class)) + geom_beeswarm(size=.5) + scale_y_discrete(expand=expansion(add=c(0.5,1)))
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-5.png" width="576" />

``` r
ggplot(mpg,aes(class, hwy)) + geom_beeswarm(size=1.1)
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-6.png" width="576" />

``` r
# With automatic dodging
ggplot(sub_mpg, aes(class, displ, color=factor(cyl))) + geom_beeswarm(dodge.width=0.5)
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-7.png" width="576" />

#### Alternative methods

``` r
df <- data.frame(
  x = "A",
  y = sample(1:100, 200, replace = TRUE)
)
ggplot(df, aes(x = x, y = y)) + geom_beeswarm(cex = 2.5, method = "swarm") + ggtitle('method = "swarm" (default)')
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-alt-1.png" width="576" />

``` r
ggplot(df, aes(x = x, y = y)) + geom_beeswarm(cex = 2.5, method = "compactswarm") + ggtitle('method = "compactswarm"')
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-alt-2.png" width="576" />

``` r
ggplot(df, aes(x = x, y = y)) + geom_beeswarm(cex = 2.5, method = "hex") + ggtitle('method = "hex"')
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-alt-3.png" width="576" />

``` r
ggplot(df, aes(x = x, y = y)) + geom_beeswarm(cex = 2.5, method = "square") + ggtitle('method = "square"')
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-alt-4.png" width="576" />

``` r
ggplot(df, aes(x = x, y = y)) + geom_beeswarm(cex = 2.5, method = "center") + ggtitle('method = "center"')
```

<img src="README_files/figure-gfm/ggplot2-beeswarm-alt-5.png" width="576" />

#### Different point distribution priority

``` r
#With different beeswarm point distribution priority
dat<-data.frame(x=rep(1:3,c(20,40,80)))
dat$y<-rnorm(nrow(dat),dat$x)
ggplot(dat,aes(x,y)) + geom_beeswarm(cex=2) + ggtitle('Default (ascending)') + scale_x_continuous(expand=expansion(add=c(0.5,.5)))
```

<img src="README_files/figure-gfm/ggplot2-priority-1.png" width="576" />

``` r
ggplot(dat,aes(x,y)) + geom_beeswarm(cex=2,priority='descending') + ggtitle('Descending') + scale_x_continuous(expand=expansion(add=c(0.5,.5)))
```

<img src="README_files/figure-gfm/ggplot2-priority-2.png" width="576" />

``` r
ggplot(dat,aes(x,y)) + geom_beeswarm(cex=2,priority='density') + ggtitle('Density') + scale_x_continuous(expand=expansion(add=c(0.5,.5)))
```

<img src="README_files/figure-gfm/ggplot2-priority-3.png" width="576" />

``` r
ggplot(dat,aes(x,y)) + geom_beeswarm(cex=2,priority='random') + ggtitle('Random') + scale_x_continuous(expand=expansion(add=c(0.5,.5)))
```

<img src="README_files/figure-gfm/ggplot2-priority-4.png" width="576" />

#### Corral runaway points

``` r
set.seed(1995)
df2 <- data.frame(
  y = rnorm(1000),
  id = sample(c("G1", "G2", "G3"), size = 1000, replace = TRUE)
)
p <- ggplot(df2, aes(x = id, y = y, colour = id))

# use corral.width to control corral width
p + geom_beeswarm(cex = 2.5, corral = "none", corral.width = 0.9) + ggtitle('corral = "none" (default)')
```

<img src="README_files/figure-gfm/ggplot2-corral-1.png" width="576" />

``` r
p + geom_beeswarm(cex = 2.5, corral = "gutter", corral.width = 0.9) + ggtitle('corral = "gutter"')
```

<img src="README_files/figure-gfm/ggplot2-corral-2.png" width="576" />

``` r
p + geom_beeswarm(cex = 2.5, corral = "wrap", corral.width = 0.9) + ggtitle('corral = "wrap"')
```

<img src="README_files/figure-gfm/ggplot2-corral-3.png" width="576" />

``` r
p + geom_beeswarm(cex = 2.5, corral = "random", corral.width = 0.9) + ggtitle('corral = "random"')
```

<img src="README_files/figure-gfm/ggplot2-corral-4.png" width="576" />

``` r
p + geom_beeswarm(cex = 2.5, corral = "omit", corral.width = 0.9) + ggtitle('corral = "omit"')
```

    ## Warning: Removed 303 rows containing missing values (geom_point).

<img src="README_files/figure-gfm/ggplot2-corral-5.png" width="576" />

------------------------------------------------------------------------

Authors: Erik Clarke, Scott Sherrill-Mix, and Charlotte Dawson


NEWS.md
---
editor_options: 
  markdown: 
    wrap: 80
---

# ggbeeswarm

## Development version

**Bugfixes:**

- `position_quasirandom()` default `dodge.width` is now `NULL` instead of 0 (#79)
- `geom_beeswarm()` now works with values of 0 on log scales, i.e. infinite
  values (#87)
- A few stray references to the deprecated `groupOnX` argument are removed.

## v0.7.1

This release incorporates all the incredible work done by @csdaw to refactor the
various geom and position functions, fix outstanding issues, and remove
dependencies on deprecated ggplot2 functions like `qplot`.

This version will be on CRAN.

**New features:**

-   Can now use new features from the `beeswarm` package, including compact
    swarms and 'corralling' points in overlapping groups.
-   Updated examples and documentation.

**Changes:**

-   The `groupOnX` argument is deprecated. `ggplot2` now has logic to determine
    when a plot is 'flipped' automatically so this argument is no longer
    necessary.
-   The `beeswarmArgs` argument is deprecated as its arguments are now called in
    `geom_beeswarm` directly.
-   `position_beeswarm` is once again an exported function.
-   Examples and documentation no longer use `ggplot2::qplot` as it has been
    hard- deprecated in `ggplot2`
-   The license has been clarified to be GPL-3, not GPL \>= 2.
-   Minimum R and `ggplot2` versions have changed to 3.5.0 and 3.3.0,
    respectively.

**Bugfixes:**

Too many to summarize here.


R/geom-beeswarm.R
#' Points, jittered to reduce overplotting using the beeswarm package
#'
#' The beeswarm geom is a convenient means to offset points within categories to
#'  reduce overplotting. Uses the beeswarm package
#'
#' @section Aesthetics:
#' \Sexpr[results=rd,stage=build]{ggplot2:::rd_aesthetics("geom", "point")}
#'
#' @inheritParams ggplot2::geom_point
#' @inheritParams offset_beeswarm
#' @param dodge.width Amount by which points from different aesthetic groups
#' will be dodged. This requires that one of the aesthetics is a factor.
#' @param groupOnX `r lifecycle::badge("deprecated")` No longer needed.
#' @param beeswarmArgs `r lifecycle::badge("deprecated")` No longer used.
#' @import ggplot2
#' @seealso
#'  [geom_quasirandom()] an alternative method,
#'  [beeswarm::swarmx()] how spacing is determined,
#'  [ggplot2::geom_point()] for regular, unjittered points,
#'  [ggplot2::geom_jitter()] for jittered points,
#'  [ggplot2::geom_boxplot()] for another way of looking at the conditional
#'     distribution of a variable
#' @export
#' @examples
#'
#' ggplot2::qplot(class, hwy, data = ggplot2::mpg, geom = "beeswarm")
#' # Generate fake data
#' distro <- data.frame(
#'   "variable" = rep(c("runif", "rnorm"), each = 100),
#'   "value" = c(runif(100, min = -3, max = 3), rnorm(100))
#' )
#' ggplot2::qplot(variable, value, data = distro, geom = "beeswarm")
#' ggplot2::ggplot(distro, aes(variable, value)) +
#'   geom_beeswarm(priority = "density", size = 2.5)
geom_beeswarm <- function(mapping = NULL,
                          data = NULL,
                          stat = "identity",
                          ...,
                          method = "swarm",
                          cex = 1,
                          side = 0L,
                          priority = "ascending",
                          fast = TRUE,
                          dodge.width = NULL,
                          corral = "none",
                          corral.width = 0.9,
                          groupOnX = NULL,
                          beeswarmArgs = list(),
                          na.rm = FALSE,
                          show.legend = NA,
                          inherit.aes = TRUE) {
  if (!missing(groupOnX)) {
    lifecycle::deprecate_soft(
      when = "0.7.1", what = "geom_beeswarm(groupOnX)",
      details = "ggplot2 now handles this case automatically."
    )
  }

  if (!missing(beeswarmArgs)) {
    lifecycle::deprecate_soft(
      when = "0.7.1", what = "geom_beeswarm(beeswarmArgs)"
    )
  }

  if (!method %in% c("swarm", "compactswarm", "hex", "square", "centre", "center")) {
    stop(sprintf("The method must be one of: swarm, compactswarm, hex, square, center, or centre."))
  }
  if (!corral %in% c("none", "gutter", "wrap", "random", "omit")) {
    stop(sprintf("The corral argument must be one of: none, gutter, wrap, random, or omit."))
  }

  position <- position_beeswarm(
    method = method,
    cex = cex,
    side = side,
    priority = priority,
    fast = fast,
    dodge.width = dodge.width,
    corral = corral,
    corral.width = corral.width
  )

  layer(
    data = data,
    mapping = mapping,
    stat = stat,
    geom = GeomPoint,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      ...
    )
  )
}
